### PR TITLE
cdc: add sink memory quota

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve?

There will be a series of PRs aims to fix the TiKV CDC OOM issue (too many pending send messages are buffered in memory).

To resolve the issue, we have two strategies,
1. Limit the speed of producing messages during the initial scan stage. #9948 
2. Set an upper bound of pending send messages. #9885 

The series of PRs adopts both strategies.

It limits the total memory used in the sink by setting a memory quota.

Plan:
- [x] take1: cherry-pick speed limiter. #10108 
- [x] take2: set an upper bound of an internal channel used in CDC #10142 
- [ ] take3: add memory quota to CDC and kill connection if it uses too much memory. <-- we are here.
- [ ] take4: cherry-pick other bugfixes in #9885 

See full changes: https://github.com/tikv/tikv/compare/master...overvenus:cdc/flowcontrol/all

Cc #8168 #9996

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- Limit CDC sink memory consumption.

### CODE REVIEW

**Comment by overvenus** [comment]
/cc @liuzix @amyangfei @ben1009

PTAL

File: components/cdc/src/channel.rs [hunk 1]
```diff
@@ -19,17 +30,64 @@ const CDC_MSG_MAX_BATCH_SIZE: usize = 128;
...
+    }
+    pub fn cap(&self) -> usize {
+        self.capacity
+    }
+    fn alloc(&self, bytes: usize) -> bool {
+        let mut total_bytes = self.in_use.load(Ordering::Relaxed);
+        loop {
+            if total_bytes + bytes > self.capacity {
+                return false;
+            }
```
Comment (amyangfei) [inline components/cdc/src/channel.rs:57]: In theory this check could have false positive, when `total_bytes + bytes > self.capacity` in L55, but the `total_bytes` has been decreased actually.

File: components/cdc/src/channel.rs [hunk 2]
```diff
@@ -125,13 +201,63 @@ impl Drain {
...
+    }
+}
+
+impl<S: Stream + Unpin> Drop for DrainGrpcMessage<S> {
+    // Must poll all pending messages to free memory quota.
+    fn drop(&mut self) {
+        // Receiver maybe dropped before senders, so set closed flag to prevent
+        // futher messages.
+        self.closed.store(true, Ordering::Release);
+        std::sync::atomic::compiler_fence(Ordering::SeqCst);
```
Comment (liuzix) [inline components/cdc/src/channel.rs:238]: Why is this necessary?

File: components/cdc/src/channel.rs [hunk 3]
```diff
@@ -125,13 +201,63 @@ impl Drain {
...
+                let item = futures::future::poll_fn(|cx| {
+                    match Pin::new(&mut next).poll(cx) {
+                        // In case it is dropped before senders.
+                        Poll::Pending => Poll::Ready(None),
+                        t => t,
+                    }
+                })
+                .await;
+                if item.is_none() {
+                    break;
```
Comment (liuzix) [inline components/cdc/src/channel.rs:252]: If `next` yields a `Poll::Pending`, will the loop break here? If so, why is it acceptable?

File: components/cdc/src/channel.rs [hunk 4]
```diff
@@ -125,13 +201,63 @@ impl Drain {
...
+        // Receiver maybe dropped before senders, so set closed flag to prevent
+        // futher messages.
+        self.closed.store(true, Ordering::Release);
+        std::sync::atomic::compiler_fence(Ordering::SeqCst);
+        let start = Instant::now();
+        let mut drain = Box::pin(async {
+            loop {
+                let mut next = self.stream.next();
+                let item = futures::future::poll_fn(|cx| {
+                    match Pin::new(&mut next).poll(cx) {
```
Comment (liuzix) [inline components/cdc/src/channel.rs:244]: Can we use `poll_unpin`?

File: components/cdc/src/service.rs [hunk 5]
```diff
@@ -356,9 +360,9 @@ impl ChangeData for Service {
         let peer = ctx.peer();
         let scheduler = self.scheduler.clone();
 
-        let mut rx = event_drain.drain_grpc_message();
         ctx.spawn(async move {
-            let res = sink.send_all(&mut rx).await;
+            // let res = sink.send_all(&mut rx).await;
```
Comment (liuzix) [inline components/cdc/src/service.rs:364]:
```suggestion
```

**Review by liuzix — APPROVED** [review]

**Review by amyangfei — APPROVED** [review]

**Comment by overvenus** [comment]
Let's track take4 in #10374 .
